library/std/src/sys/windows/process.rs

```
@@ -831,6 +865,75 @@ fn make_dirp(d: Option<&OsString>) -> io::Result<(*const u16, Vec<u16>)> {
     }
 }
 
+struct ProcThreadAttributeList(Box<[MaybeUninit<u8>]>);
+
+impl Drop for ProcThreadAttributeList {
+    fn drop(&mut self) {
+        let lp_attribute_list = self.0.as_mut_ptr() as _;
+        unsafe { c::DeleteProcThreadAttributeList(lp_attribute_list) }
+    }
+}
+
+/// Wrapper around the value data to be used as a Process Thread Attribute.
+struct ProcThreadAttributeValue {
+    data: Box<dyn Send + Sync>,
+    size: usize,
+}
+
+fn make_proc_thread_attribute_list(
+    attributes: &BTreeMap<usize, ProcThreadAttributeValue>,
+) -> io::Result<ProcThreadAttributeList> {
+    // To initialize our ProcThreadAttributeList, we need to determine
+    // how many bytes to allocate for it. The Windows API simplifies this process
+    // by allowing us to call `InitializeProcThreadAttributeList` with
+    // a null pointer to retrieve the required size.
+    let mut required_size = 0;
+    let attribute_count = attributes.len() as u32;
```

This should return an error if `attributes.len()` overflows a `u32`. Setting that many attributes is a bug so we should tell the user about it rather than silently truncating.

---

> This should return an error if `attributes.len()` overflows a `u32`

Would this be an `std::io::ErrorKind::Other`?

---

Or should the `raw_attribute` method check the length of the `BTreeMap` on every call?

---

Personally, I disagree that it should return an error. This condition is so unrealistic that the only way I could see it happening is through some arcane UB or process corruption. At that point, I feel like a panic is warranted. Even **if** someone wrote code to fill up a u32 in attributes, they would almost assuredly panic due to OOM first.

---

I Kind of agree, my first thought was to add a not in the documentation that this will happen if you supply more than `u32::Max`.

---

But I think a good compromise would be to add a note to the documentation that the number of attributes may not exceed `u32::MAX` and if it does that spawning of the process will fail. This way there is convenience when creating the attribute and the error is still handle when creating the process.

---

> Even if someone wrote code to fill up a u32 in attributes, they would almost assuredly panic due to OOM first.

I mean the `BTreeMap` would take up around 32GB without overhead but it is still possible.

---

@ChrisDenton looks good?

---

Looks good, thanks!